### PR TITLE
don't base64-encode html

### PR DESCRIPTION
**Pull request by jaspervdj**
I have an application running for the ZuriHac registration that used my own hacked-together lambda runtime -- I think `hal` didn't exist at the time.

I am in the process of porting this to `hal`, and found that I can't get plain HTML pages to render in the browser, they come out as base64-encoded versions.

Adding this Content-Type fixed it, but I wonder if we should not have the entirety of `text/` as a prefix?

### CODE REVIEW

**Comment by JackKelly-Bellroy** [comment]
> I wonder if we should not have the entirety of `text/` as a prefix?

I think you might be right. I'm also not sure why that's doing a `T.isPrefixOf` check. It might have been a misguided attempt to capture MIME types which denote responses in a particular format, but actually those are suffixed with things like `+json` (e.g., `application/problem+json`).

How does this sound: avoid base64-encoding the response when any of the following are true of the content-type:

* starts with `text/`
* is `application/json`
* ends with `+json`
* ends with `+xml`

What do you think? Also, would you mind adding a CHANGELOG.md entry?

**Comment by lrworth** [comment]
What about `application/xml`? It might be better to make the decision function user-configurable with a reasonable default, since it seems impossible for wai-handler-hal to make this decision for every possible MIME type.

**Comment by JackKelly-Bellroy** [comment]
Yes, we should support `application/xml` too, as part of the sensible default.

@jaspervdj which way do you want to take this? If you want to get your change out ASAP so you can keep building ZuriHac stuff, then we'll do that and move the broader MIME type discussion to another issue.

**Comment by jaspervdj** [comment]
> @jaspervdj which way do you want to take this? If you want to get your change out ASAP

No hurry on my part; I already pinned the commit where I added `text/html` in our project so I'm not blocked on this at all.

Making this user-configurable with a good default seems like the right direction to me.

I guess a function `Text -> Bool` is nicer than a list of prefixes and suffixes?

**Comment by JackKelly-Bellroy** [comment]
All right, then let's take in a `Text -> Bool` parameter. I think we probably want to collect the new predicate and the existing `Vault` and `PortNumber` arguments to `runWithContext` all into an options record, and provide a default value for that record. That keeps `run` as the common-case simple interface, and `runWithContext` has all the necessary knobs.

**Comment by jaspervdj** [comment]
I added this change, including a test.  Unfortunately this is a breaking change, but I guess at least the `Options` datatype can now be extended more easily...

File: wai-handler-hal.cabal
Comment (JackKelly-Bellroy) [inline wai-handler-hal.cabal:80]: If `tasty-hunit ^>=0.10` also works, can you please widen these bounds?

File: test/Network/Wai/Handler/HalTest.hs
Comment (JackKelly-Bellroy) [inline test/Network/Wai/Handler/HalTest.hs:26]: We generally format our code with `ormolu`, which will sadly wreck this nice layout. Can you please reformat? You might like to make a list of tuples if you think that makes a better layout.

File: wai-handler-hal.cabal
Comment (JackKelly-Bellroy) [inline wai-handler-hal.cabal:1]: Please update the package version and write a `CHANGELOG.md` entry — I'll do a release after we get this merged.

File: src/Network/Wai/Handler/Hal.hs
Comment (JackKelly-Bellroy) [inline src/Network/Wai/Handler/Hal.hs:122]: Can you please drop the prefix on the record selectors? I think it's unlikely we'll hit too many name clashes, and modern extensions and libraries make un-prefixed fields fairly nice to work with.

File: src/Network/Wai/Handler/Hal.hs
Comment (JackKelly-Bellroy) [inline src/Network/Wai/Handler/Hal.hs:117]: Thank you for good haddocks all through this PR.

File: src/Network/Wai/Handler/Hal.hs
Comment (JackKelly-Bellroy) [inline src/Network/Wai/Handler/Hal.hs:159]:
```suggestion
  -- | Configuration options. 'defaultOptions' provides sensible defaults.
```

File: src/Network/Wai/Handler/Hal.hs
Comment (JackKelly-Bellroy) [inline src/Network/Wai/Handler/Hal.hs:36]: I would prefer if we used the following order in the export list and when declaring these types and functions within the module:
```suggestion
    runWithContext,
    Options (..),
    defaultOptions,
```
Reason: The reader is then presented with the two main entry points to the module, and then the "what's an `Options`, and how do I make one?" is immediately raised and answered.

**Review by JackKelly-Bellroy — CHANGES_REQUESTED** [review]
Most of these are pretty trivial, and I look forward to merging and releasing it ASAP.

File: wai-handler-hal.cabal
Comment (JackKelly-Bellroy) [inline wai-handler-hal.cabal:80]:
```suggestion
    , tasty-hunit      >=0.9 && <0.11
```
(and then `cabal-fmt`)

**Review by JackKelly-Bellroy — APPROVED** [review]
Thanks very much. Please set the release date to 2023-12-16; I'll have some time tomorrow morning to push the release to Hackage. I'm happy to use GH's squash-merge feature on this PR, so don't worry about doing that.

**Comment by JackKelly-Bellroy** [comment]
Released as 0.3.0.0. Thanks for the PR.

**Pull request merged**
